### ISSUE TITLE
Sequence of partial order fills on same order is broken

### ISSUE THREAD

**Issue by bam4564** [ProjectOpenSea/seaport-js#439]
### Component

API

### Have you ensured that all of these are up to date?

- [X] seaport-js
- [X] Node (minimum v16)

### What version of seaport-js are you on?

1.0.3

### What function is the bug in?

fulfillOrder

### Operating System

macOS (Apple Silicon)

### Describe the bug

**The issue**: You cannot partially fill the same order multiple times. This issue impacts all versions of the library. 

This flow is never actually tested in `partial-fulfill.spec.ts` and it should be when implementing a fix for this issue. 

I'm going to walk through what happens when calling the function `fulfillOrder` multiple times on the same order with differing values for `unitsToFill` (signifying a partial order fill). Let's assume we have an order with the following structure 

```typescript 
// abbreviated for clarity 
const order = {
  ...
  // assume startPrice == endPrice 
  offer: [{
    token: tokenA, startPrice: 25 * 1e18
  }],
  consideration: [{
    token: tokenB, startPrice: 5 * 1e18
  }]
}
```
And that we want to execute a sequence of fills. The first for 50% of the original order amount and the second for 7% of the original order amount (order will be 57% filled upon completion of both actions). 

### Fill 50% 
1. Compute input value for `unitsToFill` 
- getMaximumSizeForOrder(order) -> findGcd([5 * 1e18, 25 * 1e18]) = 5 * 1e18 
- fractionToFill = .5 so `unitsToFill` = 5 * 1e18 * .5 * 1e18 / 1e18 = 2.5 * 1e18 
2. Call `fulfillOrder` with order struct and `unitsToFill` = 2.5 * 1e18. 
3. [Here](https://github.com/ProjectOpenSea/seaport-js/blob/cc8b571e082554672bacdad70815f619d7009090/src/seaport.ts#L853), we compute order status. This will return all zeros for queried values as no fills have previously occurred. 
4. Given that `unitsToFill` is defined, this will lead to a subsequent call to `fulfillStandardOrder`
- `unitsToFill` is passed through 
- `totalFilled` is 0 
- `totalSize` is getMaximumSizeForOrder(sanitizedOrder) which is 5 * 1e18 as detailed in step (1) 

Now let's step into `fulfillStandardOrder`
5. [Here](https://github.com/ProjectOpenSea/seaport-js/blob/cc8b571e082554672bacdad70815f619d7009090/src/utils/fulfill.ts#L369), we modify the amounts in the order by the fraction unitsToFill / totalSize = 2.5 * 1e18 / 5 * 1e18 = .5 which is 50% as expected. 
6. call getAdvancedOrderNumeratorDenominator
- maxUnits = 5 * 1e18 
- unitsToFillBn = 2.5 * 1e18 
- unitsGcd = 2.5 * 1e18 
- numerator = 1
- denominator = 2 
These values are then used to build the transaction. The numerator and denominator will be stored on chain and associated with the order hash. 

### Fill 7% 
1. Compute input value for `unitsToFill` 
- getMaximumSizeForOrder(order) -> 5 * 1e18
- fractionToFill = .07 so `unitsToFill` = 5 * 1e18 * .07 * 1e18 / 1e18 = .35 * 1e18 
2. Call `fulfillOrder` with order struct and `unitsToFill` = .35 * 1e18. 
3. [Here](https://github.com/ProjectOpenSea/seaport-js/blob/a166e9a364ee69b88484e3b9e3d2fc7ba4bebf32/src/seaport.ts#L843), we compute order status. This will return 
- `totalFilled` (numerator) = 1
- `totalSize` (denominator = 2
4. Given that `unitsToFill` is defined, this will lead to a subsequent call to `fulfillStandardOrder`
- `unitsToFill`, `totalFilled`, and `totalSize` are passed through. 
Now let's step into `fulfillStandardOrder`
5. [Here](https://github.com/ProjectOpenSea/seaport-js/blob/main/src/utils/fulfill.ts#L369), we modify the amounts in the order by the fraction unitsToFill / totalSize = .35 * 1e18 / 2 ~= 0 which is incorrect, this should be .07 since we are filling 7% of the order. 

The issue here is that we always use the fraction `unitsToFill / totalSize` to modify the order amounts. 
1. In the case where there are no prior fills, this is fine. 
2. In the case where there are prior fills, there is a scaling issue here.

The scaling issue is a result of the fact that in (1) the value of `totalSize` is the output of calling `getMaximumSizeForOrder` while in case (2) the value of `totalSize` is read from on chain (and is normalized). 

This is a limitation of the broken API and multiple fills are supported at the seaport protocol level (i.e. fill 50%, then 7%, then 43% is possible). I am positive of this fact because I have a working example of this in production. I was able to come up with a hacky workaround to get sequential partial order fills working in seaport-js 1.0.3 due to another bug that was patched in version 2.0.0 (relevant PR: https://github.com/ProjectOpenSea/seaport-js/pull/250). This workaround is only possible because there used to be an issue where the order hash computed on chain differred from the seaport-js internal implementation of `getOrderStatus` when the salt was passed in as a stringified bigint in non-hexadecimal form. Given that the order hash computed internally didn't match the on chain hash, prior fills were not picked up leading to values of `totalFilled` and `totalSize` being 0 even though prior fills existed. Even though this was erroneous, the modified order based on the fraction `unitsToFill` / `totalSize` in this case was right so I was able to make things work. 

I opted to write out the state transitions rather than try to create a minimum reproducible repo as I'm a bit short on bandwidth right now, but I hope this writeup is clear enough to follow and understand the issue. This issue would be pretty simple to validate by updating the partial order fills test to actually try a sequence of partial fills.

**Comment by bam4564** [comment]
Bumping this!

**Comment by ryanio** [comment]
thanks for the report, if you were interested in opening a PR to add the suggested test and fix, it would be much appreciated! otherwise we can try to address soon

**Comment by bam4564** [comment]
Will let you know if I or my team gets some bandwidth to help out here. Currently we're quite busy but I'll revisit this later.

**Comment by ryanio** [comment]
thanks again for the detailed report, i'll work on a PR to fix

**Comment by ryanio** [comment]
Hey @bam4564, here is my PR to fix: #490 if you don't mind taking a look and see if it fixes the issue.

I added tests to partial fill twice to recreate the unitsToFill issue, in both fulfillStandardOrder and fulfillAvailableOrders.

**Comment by bam4564** [comment]
@ryanio thanks so much for the work here. A little busy this week with work launching a few things that require stability in our systems. Will pass this off to our frontend team and they can experiment with this new version so see if it solves our issues. Should be able to test sometime this upcoming week!
